engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/KvmNonManagedStorageDataMotionStrategy.java

```diff
@@ -195,7 +195,10 @@ protected boolean shouldMigrateVolume(StoragePoolVO sourceStoragePool, Host dest
     @Override
     protected void copyTemplateToTargetFilesystemStorageIfNeeded(VolumeInfo srcVolumeInfo, StoragePool srcStoragePool, DataStore destDataStore, StoragePool destStoragePool,
             Host destHost) {
-        VMTemplateStoragePoolVO sourceVolumeTemplateStoragePoolVO = vmTemplatePoolDao.findByPoolTemplate(destStoragePool.getId(), srcVolumeInfo.getTemplateId(), null);
+        VMTemplateStoragePoolVO sourceVolumeTemplateStoragePoolVO = null;
+        if (srcVolumeInfo != null && srcVolumeInfo.getTemplateId() != null) {
+            sourceVolumeTemplateStoragePoolVO = vmTemplatePoolDao.findByPoolTemplate(destStoragePool.getId(), srcVolumeInfo.getTemplateId(), null);
+        }
         if (sourceVolumeTemplateStoragePoolVO == null && destStoragePool.getPoolType() == StoragePoolType.Filesystem) {
             DataStore sourceTemplateDataStore = dataStoreManagerImpl.getRandomImageStore(srcVolumeInfo.getDataCenterId());
```

Comment:
@Pearl1594 can you help me understanding on which object we are getting NPE. I'm seeing only srcVolumeInfo as possible object but it would have failed earlier at 198 Line.

Reply:
@harikrishna-patnala I've updated the exception,  you're right, the exception happens at line198 - I had added a few debug lines which was why it showed the exception occuring at L200

Reply:
@harikrishna-patnala The problem is that `findByPoolTemplate` receives a `long` (primitive type) to `template` and `getTemplateId` returns a `Long` (object type). Java will do an autounboxing, but as `getTemplateId` returns `null` for non-template volume (data disks), it will generate a NPE when trying to autounbox this value.

Reply:
@GutoVeronezi wouldn't `srcVolumeInfo.getTemplateId() != null` prevents what you mentioned?

Reply:
@GabrielBrascher It will. I mention it to explain why it is happening.